### PR TITLE
Don't trim highlight snippet if number_of_fragments is 0 #101803

### LINKED ISSUE #101803: Inconsistent Text Trimming in `Unified` Highlighter with `number_of_fragments` Set to `0`
### Elasticsearch Version

8.10.4

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

Elastic Cloud - GCP - Iowa (us-central1)

### Problem Description

When using the Elasticsearch `unified` highlighter, I've encountered an inconsistency related to the `number_of_fragments` option. The issue arises when `number_of_fragments` is set to 0, indicating that we want to highlight the entire content of a field. However, the `unified` highlighter unexpectedly trims the text.

The expected highlight for the input text `" a b c d "` with `number_of_fragments` set to `0` should be:

    "highlight" : {
      "text" : [
        " a  b (c) d "
      ]
    }

However, the actual response is:

    "highlight" : {
      "text" : [
        "a  b (c) d"
      ]
    }

### Steps to Reproduce

    PUT test_trim
    {
      "mappings": {
        "properties": {
          "text": {
            "type": "text",
            "analyzer": "whitespace"
          }
        }
      }
    }

    PUT test_trim/_doc/1
    {
      "text": " a  b c d "
    }

    GET test_trim/_search
    {
      "query": {
        "match": {
          "text": "c"
        }
      },
      "highlight": {
        "pre_tags": "(", 
        "post_tags": ")", 
        "fields": {
          "text": {}
        },
        "number_of_fragments": 0, 
        "type": "unified"
      }
    }

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by yashdamani**
Don't trim highlight snippet if number_of_fragments is 0

Closes  #101803

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/lucene/search/uhighlight/CustomPassageFormatterTests.java [hunk 1]
```diff
@@ -21,7 +21,7 @@ public class CustomPassageFormatterTests extends ESTestCase {
     public void testSimpleFormat() {
         String content = "This is a really cool highlighter. Unified highlighter gives nice snippets back. No matches here.";
 
-        CustomPassageFormatter passageFormatter = new CustomPassageFormatter("<em>", "</em>", new DefaultEncoder());
+        CustomPassageFormatter passageFormatter = new CustomPassageFormatter("<em>", "</em>", new DefaultEncoder(), 0);
```
Comment (AlexHelmutSonntag) [inline server/src/test/java/org/elasticsearch/lucene/search/uhighlight/CustomPassageFormatterTests.java:24]: Does it make sense to add context in the Test Class to this new argument used ?

Comment (yashdamani) [reply server/src/test/java/org/elasticsearch/lucene/search/uhighlight/CustomPassageFormatterTests.java:24]: I could, but wouldn't going the class `CustomPassageFormatter` provide that context? I'm assuming you're suggesting adding context that the new parameter is for `numberOfFragments`? Correct me if I am wrong in my understanding your suggestion.

**Comment by yashdamani** [comment]
@mayya-sharipova would you be able to help with reviewing this PR?

**Review by mayya-sharipova — CHANGES_REQUESTED** [review]
@yashdamani Thank you for your contribution. The approach makes sense to me.


Please add rest tests into `rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.highlight` for the change. Here is the guidance about [YAML rest tests](https://github.com/elastic/elasticsearch/blob/main/TESTING.asciidoc#testing-the-rest-layer).

**Comment by yashdamani** [comment]
@mayya-sharipova I've added the rest test file for the change. I'm not sure the about preceding number in the naming convention for the .yml files. I've added the case for number_of_fragments > 0 and number_of_fragments = 0. Is a default case (where the param number_of_fragments isn't mentioned) also required? Also, I've used the same input as the original issue linked with this PR.

**Comment by mayya-sharipova** [comment]
@elasticsearchmachine test this please

File: docs/changelog/106306.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+pr: 99961
+summary: "added fix for inconsistent text trimming in Unified Highlighter"
+area: Search
```
Comment (mayya-sharipova) [inline docs/changelog/106306.yaml:3]: please change to :  `area: Highlighting`

**Comment by yashdamani** [comment]
@mayya-sharipova I ran `./gradlew check` earlier in the day, and it ran for almost an hour, and I ended up exiting it. Is that expected?

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/510_fragment_trimming_fix.yml [hunk 3]
```diff
@@ -0,0 +1,59 @@
+setup:
```
[the remark this answers is not in the document]
Comment (yashdamani) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/510_fragment_trimming_fix.yml:1]: Thank you! Made all the changes as suggested.

**Comment by mayya-sharipova** [comment]
@yashdamani Indeed `./gradlew check` runs long, we usually don't run it locally. We run `./gradlew precommit` and then applicable tests.

**Comment by mayya-sharipova** [comment]
@elasticsearchmachine test this please

**Comment by mayya-sharipova** [comment]
@yashdamani Are you are able to access CI build failures and see and correct tests that are failing? 

```
 ./gradlew ':server:test' --tests "org.elasticsearch.lucene.search.uhighlight.CustomUnifiedHighlighterTests.testRepeatPhrase" -Dtests.seed=512E5E4D03C679D1 -Dtests.locale=de-LU -Dtests.timezone=America/Argentina/Buenos_Aires -Druntime.java=21

./gradlew ':server:test' --tests "org.elasticsearch.lucene.search.uhighlight.CustomUnifiedHighlighterTests.testSmallSentenceBoundedBreakIterator" -Dtests.seed=512E5E4D03C679D1 -Dtests.locale=de-LU -Dtests.timezone=America/Argentina/Buenos_Aires -Druntime.java=21
 

<br class="Apple-interchange-newline">
```

**Comment by yashdamani** [comment]
@mayya-sharipova I can't access the builds. On clicking Details it asks me to sign in to buildkite.

**Comment by mayya-sharipova** [comment]
@yashdamani The test failures are only in `CustomUnifiedHighlighterTests`, so you can run:

 ` ./gradlew ':server:test' --tests "org.elasticsearch.lucene.search.uhighlight.CustomUnifiedHighlighterTests.*`

**Review by yashdamani — COMMENTED** [review]
@mayya-sharipova there were errors in `CustomPassageFormatterTests` as well. I've made changes to now include tests for different values of `numberOfFragments`.  Would [this](https://github.com/elastic/elasticsearch/pull/106306/commits/646ca23fda9ef3e6cd01e791885d770d464932d0) be the right approach to modify the tests? If yes, then I'll use the same approach for `CustomUnifiedHighlighterTests` as well, which failed in the build.

**Comment by mayya-sharipova** [comment]
@yashdamani Please don't force push changes while PR is under review, as it is difficult to see the changes from the previous review.

**Comment by yashdamani** [comment]
@mayya-sharipova alright, will keep that in mind going ahead.

File: server/src/test/java/org/elasticsearch/lucene/search/uhighlight/CustomPassageFormatterTests.java [hunk 4]
```diff
@@ -21,7 +21,10 @@ public class CustomPassageFormatterTests extends ESTestCase {
     public void testSimpleFormat() {
         String content = "This is a really cool highlighter. Unified highlighter gives nice snippets back. No matches here.";
 
-        CustomPassageFormatter passageFormatter = new CustomPassageFormatter("<em>", "</em>", new DefaultEncoder());
```
[the remark this answers is not in the document]
Comment (yashdamani) [reply server/src/test/java/org/elasticsearch/lucene/search/uhighlight/CustomPassageFormatterTests.java:24]: Got it. I was under the impression that both the testcases (where `numberOfFragments` is 0 and > 0) need to be written.

**Review by mayya-sharipova — APPROVED** [review]
@yashdamani Thanks for your great work! Merging for the 8.14 release.

**Comment by yashdamani** [comment]
@mayya-sharipova Thank you!

**Pull request merged**